<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="UTF-8">
<title>SeqWare › SeqWare Query Engine</title>
<meta content="nanoc 3.4.0" name="generator">
<link href="../../assets/style-v17.css" media="screen" rel="stylesheet" type="text/css">
<link href="http://feeds.feedburner.com/seqware" rel="alternate" title="The SeqWare Blog" type="application/atom+xml">
<script type="text/javascript"> var _gaq = _gaq || []; _gaq.push(['_setAccount', 'UA-34523087-1']); _gaq.push(['_trackPageview']); (function() { var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true; ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js'; var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s); })(); </script>
</head>
<body>
<div class="wrapper">
<!-- ***** navigation ***** -->
<div class="nav">
<ol>
<li><a href="../../"><span>Home</span></a></li>
<li><a href="../../blog/"><span>News</span></a></li>
<li><a href="../"><span>Documentation</span></a></li>
<li><a href="../../community/"><span>Community</span></a></li>
<li><a href="../../partners/"><span>Partners</span></a></li>
<li><a href="../../about/"><span>About</span></a></li>
</ol>
</div>
<!-- ***** sidebar ***** -->
<div class="side">
<ol>
<li><a href="../1-introduction/"><span>Introduction to SeqWare</span></a></li>
<li><a href="../2-installation/"><span>Installation</span></a></li>
<li><a href="../3-getting-started/"><span>Getting Started</span></a></li>
<li><a href="../4-metadb/"><span>SeqWare MetaDB</span></a></li>
<li><a href="../5-portal/"><span>SeqWare Portal</span></a></li>
<li><a href="../6-pipeline/"><span>SeqWare Pipeline</span></a></li>
<li><a href="../7-web-service/"><span>SeqWare Web Service</span></a></li>
<li>
<span class="active" title="You're here."><span>SeqWare Query Engine</span></span><ol class="toc">
<li><a href="#overview">Overview</a></li>
<li><a href="#admin-setup">Admin Setup</a></li>
<li><a href="#features">Features</a></li>
<li><a href="#loading-data">Loading Data</a></li>
<li><a href="#querying-data">Querying Data</a></li>
<li><a href="#making-custom-plugins">Making Custom Plugins</a></li>
<li><a href="#running-tests">Running Tests</a></li>
<li><a href="#more-to-come">More to Come &#8230;</a></li>
</ol>
</li>
<li><a href="../9-glossary/"><span>Glossary</span></a></li>
<li><a href="../10-faq/"><span>Frequently Asked Questions</span></a></li>
<li><a href="../11-api/"><span>APIs</span></a></li>
</ol>
</div>
<!-- ***** body ***** -->
<div class="body">
<div class="article">
<div class="header">
<h1>SeqWare Query Engine</h1>
</div>
<div class="section" id="overview">
<h2>Overview</h2>

<p>The SeqWare Query Engine is intended to be a universal store for sequence variants. The current iteration of the SeqWare Query engine has the ability to store and search data stored in a modern NoSQL database, HBase.</p>

</div>
<div class="section" id="admin-setup">
<h2>Admin Setup</h2>

<p>We can start by getting your development environment setup with the appropriate prerequisites. The <a href="../2-installation/">Installing with a Local VM</a> guide will give you access to a VM which has these setup correctly. However, if you wish to set this up yourself and you have git and mvn installed:</p>

<ol>
<li>You will want to go to the <a href="https://github.com/SeqWare/seqware">SeqWare Github repository</a> and run the command <code>git clone git@github.com:SeqWare/seqware.git seqware_github</code>
</li>
<li>You will also need to setup Hadoop and HBase if it is not already setup. We highly recommend <a href="https://ccp.cloudera.com/display/CDH4DOC/CDH4+Quick+Start+Guide">Cloudea&#8217;s CDH packages</a> since they are tested for package incompatibilities between Hadoop projects (which are still common). It is also worth double-checking the web interface for HBase which is usually at <a href="http://localhost:60010/master-status">http://localhost:60010/master-status</a>.</li>
</ol>
<p>For both the VM and a local development environment, continue here:</p>

<p>There are a number of constants that may have to be set, particularly for a developer. These are currently in the <code>com.github.seqware.queryengine.Constants</code> file although they will be moved to an external configuration file. In particular, you should set your <code>NAMESPACE</code> to avoid collisions with other developers and if you wish for your distribution jar to be automatically copied to the cluster when launching MapReduce tasks, you will need to correct the <code>DEVELOPMENT_DEPENDENCY</code>.</p>

<p class="warning"><strong>Note:</strong>
       It is important that you check your <code>NAMESPACE</code>, <code>HBASE_REMOTE_TESTING</code>, and <code>HBASE_PROPERTIES</code> variables. While these will be moved out to an external config file, they currently control the prefix for your tables, whether you connect to a local install of HBase, and which remote install of HBase you want to connect to respectively.
</p>


<ol>
<li>Refresh the seqware engine by doing a <code>git fetch</code> and <code>git pull</code> in the seqware_github directory.</li>
<li>If the <a href="http://localhost:60010/master-status">web interface</a> for HBase stalls or is inactive, you may need to restart the HBase processes. This can be done by the following commands:
<pre title="Title of the snippet">
sudo bash
/etc/init.d/hbase-regionserver stop
/etc/init.d/hbase-master stop
/etc/init.d/hadoop-zookeeper-server stop
/etc/init.d/hbase-regionserver start
/etc/init.d/hbase-master start
/etc/init.d/zookeeper-server start
jps
</pre>
</li>
<li>When setup of Hadoop and HBase is complete, you can go into the query-engine directory, compile it, and run the tests. Please note that the web-service and legacy directories in the root have additional dependencies and may not necessarily compile following only these instructions.
<pre title="Title of the snippet">
mvn clean install
mvn javadoc:javadoc
mvn javadoc:test-javadoc
</pre>
</li>
</ol>
</div>
<div class="section" id="features">
<h2>Features</h2>

</div>
<div class="section" id="loading-data">
<h2>Loading Data</h2>

<p>We currently load data via our command-line programs. In order to do this, you will want to go to the root directory and compile a full version of our jar with dependencies included. You will also need to skip the tests for our other components. This should look something like:</p>

<pre title="Title of the snippet">
  <kbd>cd ...</kbd>
  <kbd>mvn install -Dmaven.test.skip=true</kbd>
  <kbd>cd seqware-distribution/target</kbd>
</pre>


<p>In this distribution directory, you can run our command line tools for import.</p>

<p>The first time you run these tools in a new namespace, you may wish to create a common reference and then create an ad hoc tag set for all terms that do not match Sequence Ontology (SO) terms.</p>

<pre title="Title of the snippet">
<span class="prompt">~/seqware_github/seqware-distribution/target$</span> <kbd>java -classpath seqware-queryengine-0.12.0-full.jar com.github.seqware.queryengine.system.ReferenceCreator</kbd>
Only 0 arguments found
ReferenceCreator <reference_name> [output_file]
<span class="prompt">~/seqware_github/seqware-distribution/target$</span> <kbd>java -classpath seqware-queryengine-0.12.0-full.jar com.github.seqware.queryengine.system.ReferenceCreator hg_19 keyValue_ref.out</kbd>
Reference written with an ID of:
59a1aca5-4a5f-4006-b395-ca8cb5dd8c50
<span class="prompt">~/seqware_github/seqware-distribution/target$</span> <kbd>cat keyValue_ref.out</kbd>
referenceID    59a1aca5-4a5f-4006-b395-ca8cb5dd8c50
<span class="prompt">~/seqware_github/seqware-distribution/target$</span> <kbd>java -classpath seqware-queryengine-0.12.0-full.jar com.github.seqware.queryengine.system.TagSetCreator</kbd>
Only 0 arguments found
TagSetCreator <tagset name> [output_file]
<span class="prompt">~/seqware_github/seqware-distribution/target$</span> <kbd>java -classpath seqware-queryengine-0.12.0-full.jar com.github.seqware.queryengine.system.TagSetCreator ad_hoc keyValue_adHoc.out</kbd>
TagSet written with an ID of:
84ec5dfb-1cba-4b7b-b06a-eebbfc3dfd60
<span class="prompt">~/seqware_github/seqware-distribution/target$</span> <kbd>cat keyValue_adHoc.out</kbd>
TagSetID    84ec5dfb-1cba-4b7b-b06a-eebbfc3dfd60
</tagset></reference_name></pre>


<p>You may also wish to pre-populate the database with SO terms in a TagSet:</p>

<pre title="Title of the snippet">
<span class="prompt">~/seqware_github/seqware-distribution/target$</span> <kbd>java -classpath seqware-queryengine-0.12.0-full.jar com.github.seqware.queryengine.system.importers.OBOImporter</kbd>
Only 0 arguments found
OBOImporter <input_file> [output_file]
<span class="prompt">~/seqware_github/seqware-distribution/target$</span> <kbd>java -classpath seqware-queryengine-0.12.0-full.jar com.github.seqware.queryengine.system.importers.OBOImporter ../../seqware-queryengine/src/test/resources/com/github/seqware/queryengine/system/so.obo keyValueOBO.out</kbd>
3870 terms written to a TagSet written with an ID of:
5e7a2327-08ac-4455-9687-ec4c3737074f
<span class="prompt">dyuen@odl-dyuen:~/seqware_github/seqware-distribution/target$</span> <kbd>cat keyValueOBO.out</kbd>
TagSetID    5e7a2327-08ac-4455-9687-ec4c3737074f
</input_file></pre>


<p>The previous steps should only really be done once when first setting up a namespace. Afterwards, the VCF file importer can be called.</p>

<pre title="Title of the snippet">
<span class="prompt">~/seqware_github/seqware-distribution/target$</span> <kbd> java -classpath seqware-queryengine-0.12.0-full.jar com.github.seqware.queryengine.system.importers.SOFeatureImporter ../../seqware-queryengine/src/test/resources/com/github/seqware/queryengine/system/FeatureImporter/consequences_annotated.vcf -o keyValueVCF.out -r 59a1aca5-4a5f-4006-b395-ca8cb5dd8c50 -s 5e7a2327-08ac-4455-9687-ec4c3737074f -w VCFVariantImportWorker </kbd>
usage: SOFeatureImporter
 -a <adhoctagset>   (optional) an ID for an ad hoc TagSet, Tags will
                    either be found or added to this set, a new TagSet
                    will be generated if none is specified here
 -c                 (optional) whether we are working with compressed
                    input
 -i <inputfile>     (required) comma separated input files
 -o <outputfile>    (optional) output file with our resulting key values
 -r <reference>     (required) the reference ID to attach our FeatureSet
                    to
 -s <tagset>        (optional) comma separated TagSet IDs, new Tags will
                    be linked to the first set that they appear, these
                    TagSets will not be modified
 -t <threads>       (optional: default 1) the number of threads to use in
                    our import
 -w <worker>        (required) the work module and thus the type of file
                    we are working with
<span class="prompt">~/seqware_github/seqware-distribution/target$</span> <kbd> java -classpath seqware-queryengine-0.12.0-full.jar com.github.seqware.queryengine.system.importers.SOFeatureImporter -i ../../seqware-queryengine/src/test/resources/com/github/seqware/queryengine/system/FeatureImporter/consequences_annotated.vcf -o keyValueVCF.out -r 59a1aca5-4a5f-4006-b395-ca8cb5dd8c50 -s 5e7a2327-08ac-4455-9687-ec4c3737074f -a 84ec5dfb-1cba-4b7b-b06a-eebbfc3dfd60 -w VCFVariantImportWorker </kbd>
FeatureSet written with an ID of:
99d3cc0e-26e0-4b23-9ebf-90fbe07a6c5e
<span class="prompt">~/seqware_github/seqware-distribution/target$</span> <kbd> echo $? </kbd>
0
<span class="prompt">~/seqware_github/seqware-distribution/target$</span> <kbd> cat keyValueVCF.out </kbd>
FeatureSetID    99d3cc0e-26e0-4b23-9ebf-90fbe07a6c5e
</worker></threads></tagset></reference></outputfile></inputfile></adhoctagset></pre>


</div>
<div class="section" id="querying-data">
<h2>Querying Data</h2>

</div>
<div class="section" id="making-custom-plugins">
<h2>Making Custom Plugins</h2>

</div>
<div class="section" id="running-tests">
<h2>Running Tests</h2>

</div>
<div class="section" id="more-to-come">
<h2>More to Come &#8230;</h2>

<p>TODO: In the future, we will be developing:</p>

<ol>
<li>REST API: We will be developing a RESTful API to allow for the development of web applications and to allow users to query and view their data easily.</li>
<li>Web App: A sample web application demoing genome browsing</li>
</ol>
</div>
</div>
</div>
</div>
<!-- ***** footer ***** -->
<div id="footer">
<p>SeqWare © 2007–2012 Brian O&#8217;Connor. SeqWare is released under the a <a rel="license" href="http://www.gnu.org/licenses/licenses.html">GNU GPL v3</a>. This site is built using the excellent <a href="http://nanoc.stoneship.org/">nanoc</a> tool and example site along with the <a href="http://www.fonts.info/info/press/free-fonts-for-font-face-embedding.htm">Graublau</a> and <a href="http://scripts.sil.org/Gentium">Gentium</a> fonts.</p>
</div>
</body>
</html>
